platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,181 @@
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state.service;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.state.primitives.ProtoBytes;
+import com.hedera.hapi.node.state.roster.Roster;
+import com.hedera.hapi.node.state.roster.RosterState;
+import com.hedera.hapi.node.state.roster.RosterState.Builder;
+import com.hedera.hapi.node.state.roster.RoundRosterPair;
+import com.hedera.hapi.platform.state.PlatformState;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.RosterStateId;
+import com.swirlds.platform.roster.RosterUtils;
+import com.swirlds.platform.roster.RosterValidator;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Read-write implementation for accessing rosters states.
+ */
+public class WritableRosterStore extends ReadableRosterStoreImpl {
+
+    /**
+     * The maximum number of active rosters to keep in the roster state.
+     */
+    public static final int MAXIMUM_ROSTER_HISTORY_SIZE = 2;
+
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        super(writableStates);
+        requireNonNull(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Sets the candidate roster in state.
+     * Setting the candidate roster indicates that this roster should be adopted as the active roster when required.
+     *
+     * @param candidateRoster a candidate roster to set. It must be a valid roster.
+     */
+    public void putCandidateRoster(@NonNull final Roster candidateRoster) {
+        requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        final Bytes incomingCandidateRosterHash =
+                RosterUtils.hash(candidateRoster).getBytes();
+
+        // update the roster state/map
+        final RosterState previousRosterState = rosterStateOrThrow();
```

Comment:
@edward-swirldslabs : given this `OrThrow` here, perhaps unchecked exceptions are okay? If not, then why isn't this one a checked exception? Also, what about all the NPEs that could be thrown from null checks, or the `IllegalStateException` thrown below?

Reply:
As long as a Schema is present that registers this service, the state _should_ not be null.